### ISSUE TITLE
Cognito Resource Replaces instead of Updating

### ISSUE THREAD

**Issue by lorecrafting** [pulumi/pulumi#2759]
When updating an attribute for Cognito resource, it deletes the User Pool and recreates a new one instead of keeping the same user pool and updating the attribute.  

This could potentially be dangerous when there are live users in the pool and it destroys the entire pool and recreates a new one.

@lukehoban has looked into this issue and apparently this is due to the Cognito API and happens in Terraform and Cloudformation also.  Luke looked into it awhile back and there may be a fix for it.

Thanks @lukehoban !!

**Comment by Davis10733** [comment]
I have same issue. After some research I found this.

https://forums.aws.amazon.com/thread.jspa?messageID=794788

According to the reply of AWS, after creating a cognito user pool, we can't edit attributes of it.

**Comment by danielrbradley** [comment]
Been having a similar issue, may or may not be related: I ran `pulumi preview --diff` and noticed the following even though I'd not made changes:
![image](https://user-images.githubusercontent.com/331676/67965149-64ea0e80-fbf9-11e9-8653-089de3969e7e.png)
Adding the following to my schema caused the diff to be empty:
```ts
      stringAttributeConstraints: {
        maxLength: "2048",
        minLength: "0",
      }
```

I don't think this will avoid a replace instead of update if you've changed other properties, but it should fix if it's triggering a recreate on every deploy.

**Comment by mk0a1a** [comment]
This also happens even without changes to the UserPool attributes. Looks like this is an issue on the underlying [Terraform provider](https://github.com/terraform-providers/terraform-provider-aws/issues/3891). The [workaround](https://github.com/terraform-providers/terraform-provider-aws/issues/3891#issuecomment-380256678-permalink) is to add `stringAttributeConstraints` to the `schema` config. Similar to @danielrbradley's comment.

The issue at Terraform has been open since March of 2018 and to me, this is insane. Recreating a UserPool causes the users to be deleted which would be deadly to deploy on a production stack. Even if the users are backed up, you'll have to send password resets to all users.

On a related note, Cognito is not the only resource I've found that does this. ACM certificates also recreate even with no changes. Here is an [open issue](https://github.com/terraform-providers/terraform-provider-aws/issues/8531) on the Terraform provider. This causes other resources that depend on an ACM certificate to update, for example a Cloudfront distribution which takes around 30 minutes to apply changes causing long wait times for minor changes.

**Comment by Superjo149** [comment]
I'm experiencing the same. Since I have multiple attributes in the schema, for me there is also an issue where the array of schemas being compared is in a different order.

**Comment by pgavlin** [comment]
Looks like this issue was fixed upstream: https://github.com/hashicorp/terraform-provider-aws/issues/8531#issuecomment-666878609. These fixes were released with version 3.0.0 of the Pulumi AWS Provider: https://github.com/pulumi/pulumi-aws/blob/master/CHANGELOG.md#300-2020-08-11.
